cc/mcts_node.cc

```diff
@@ -41,11 +83,24 @@ MctsNode::MctsNode(MctsNode* parent, Coord move)
       move(move),
       position(parent->position) {
   position.PlayMove(move);
-  // TODO(tommadams): Only call IsMoveLegal if we want to select a leaf for
-  // expansion.
-  for (int i = 0; i < kNumMoves; ++i) {
-    illegal_moves[i] = !position.IsMoveLegal(i);
+
+  // Insert a cache of ancestor Zobrist hashes at regular depths in the tree.
```

Comment:
A cache's cost grows linearly with position.n(); while a cache's benefit is equal to [number of times any child of this node is visited] * [distance between this cache and the previous cache].

So from an optimality perspective, it seems like caches should become less frequent as the game goes on longer. And there's probably some heuristic, like, "if a node has been visited X times and we're on readout 160 of 1600, then, we can probably guess that this node will be visited 10X times by the time tree search completes", and then dynamically insert a cache based on that estimate.